### PR TITLE
[usbdev] Fixes for I/O modes and expand their tests

### PR DESCRIPTION
**Pull request by mdhayter**
The differental mode reception and conversion was being done in the iomux
and potentially not passing all information to the usb_fs_rx module. Reworked
to pass the single ended and differential data to usb_fs_rx, recover them
both and use the configured mode to forward appropriately.

This changes the usbdev_iomux to only synchronize and pinflip I/O and
pass the rx d, d+ and d- to receive logic.

As part of debug fixed verilator UNOPTFLAT complaint and added a link
state representing the link being active but never seen SOF because this
is where a flipped link gets stuck between FPGA board and Linux. (The
flipped pullup will make it look like a LS not FS link so the 1.5Mbit
chatter is enough to get rx PID erros but also keeps resetting the
link so host_lost never fires).

Update verilator top and usbdpi to manage all the I/Os. And add code
to the usbdpi to be able to simulate pinflip and differential as well
as the normal pin mode. Note that because there is no other signal it
uses the tx_mode_se signal to control use of differential mode in both
tx and rx directions (so it only works if both phy config bits are the
same). It uses the DN PULLUP being asserted to detect flip.

Update hello_usbdev.c to set the phy config from gpio pins. These
are switches on the nexysvideo and can be set using gpio-dpi in
verilator.

Update top_earlgrey_nexysvideo to be able to fake swapping of USB pins.
Uses DN pullup to cause the pins used for DP and DN to be swapped, allows
testing of both flip options with one FPGA build.

Tested with verilator that all four settings of PINFLIP and DIFFIO in
hello_usb.c get the same expected output. Added a script to run all
four simulations (setting the GPIOs to control).

Tested with FPGA that PINFLIP=0,1 both work by setting the switch.

Included changes requested in software and hardware reviews.

Fixes 2598

Signed-off-by: Mark Hayter <mark.hayter@gmail.com>

### CODE REVIEW

**Comment by mdhayter** [comment]
This is the rebase of #2753 

@stefanlippuner I had trouble adding you to the review list, hope the reference here works.

File: sw/device/tests/usbdev_test.c
Comment (mcy) [inline sw/device/tests/usbdev_test.c:92]: Consider writing `/*pinflip=*/ false` and similar in such function calls, to make which parameter is which clearer.

File: sw/device/lib/usbdev.c
Comment (mcy) [inline sw/device/lib/usbdev.c:272]: Don't predeclare if you're going to assign to it immediately.

File: sw/device/examples/hello_usbdev/hello_usbdev.c
Comment (mcy) [inline sw/device/examples/hello_usbdev/hello_usbdev.c:88]: This can be `const char *`.

File: sw/device/examples/hello_usbdev/hello_usbdev.c
Comment (mcy) [inline sw/device/examples/hello_usbdev/hello_usbdev.c:137]: LOG macros already emit an implicit newline.

File: sw/device/examples/hello_usbdev/hello_usbdev.c
Comment (mcy) [inline sw/device/examples/hello_usbdev/hello_usbdev.c:140]: Note that `usbdev_init` accepts `bool`, which should be 0 or 1, but these AND operations may produce other values. You should remember to write `(gpio & kPinflipMask) != 0` to avoid this.

**Review by mcy — CHANGES_REQUESTED** [review]
I included comments in HW C as well as SW C, but only the SW comments are required.

Also, I think you might need to clang-format.

File: sw/device/examples/hello_usbdev/hello_usbdev.c
Comment (mdhayter) [reply sw/device/examples/hello_usbdev/hello_usbdev.c:140]: Shouldn't the compiler do this when it does the casy from uint32 to bool?
In any event I will move to setting in variables to avoid the repeated enumerations.

File: sw/device/tests/usbdev_test.c
Comment (mdhayter) [reply sw/device/tests/usbdev_test.c:92]: Done

File: sw/device/examples/hello_usbdev/hello_usbdev.c
Comment (mdhayter) [reply sw/device/examples/hello_usbdev/hello_usbdev.c:137]: Done

File: sw/device/examples/hello_usbdev/hello_usbdev.c
Comment (mdhayter) [reply sw/device/examples/hello_usbdev/hello_usbdev.c:140]: Note the LOG output of the new code confirms that the new bool differential is 1 (not kDiffMask==2)

**Review by mdhayter — COMMENTED** [review]
Fixes for @mcy comments and will push amended code

**Review by vogelpi — COMMENTED** [review]
Thanks @mdhayter for this PR. I overread one of your comments and accidentally reviewed the old one (which also contains the latest changes but not the rebase). I have left my comments there. Sorry for the inconvenience.

**Comment by mdhayter** [comment]
@vogelpi I'm fine with comments in the other side, although I will address them here. I wanted people to be able to see the changes there (without distractions from the rebase).

**Comment by mdhayter** [comment]
Did @vogelpi fixes recommended in #2753 and rebased again.

Runs for me on Nexys Video with both pin flips and verilator works.

TODO: Extend script with expected output checking

I suspect usbdev needs to be run through the verilog and C code format tools, I didn't find good instructions in our docs for that yet but will look.

**Comment by mdhayter** [comment]
I think I got git clang-format to fix the C code (I found a hint at the bottom of the style guide)

Is there something similar for the verilog?

**Review by vogelpi — APPROVED** [review]
Thanks @mdhayter for addressing all my comments and all the effort you put into this (and testing it on the FPGA)! From my side, there is just one minor nit but I can give you green light.

Before merging this, we also need green light from @mcy . Could you please have another look Michael?

As for the Verilog formatting: I think most people don't use auto-formatting but just use lint. You can run lint using
```
util/dvsim/dvsim.py hw/top_earlgrey/lint/top_earlgrey_lint_cfgs.hjson --tool veriblelint
```
or instead of `veriblelint` you can use `verilator`. This will lint the whole design and take a couple of minutes but I guess because `usbdev` also depends on `usb_fs_nb_pe` it is the right thing to do.

**Comment by vogelpi** [comment]
Ah and by the way, you can replace the line
```
Fixes 2598
```
in your commit message by
```
Fixes lowRISC/OpenTitan#2598.
```
Then the commit will automatically close the associated issue upon merge.

File: sw/device/examples/hello_usbdev/hello_usbdev.c
Comment (lenary) [reply sw/device/examples/hello_usbdev/hello_usbdev.c]: We've tried to avoid implicit int to bool casts in other code so far, and instead favour explicitly writing out the comparison that is intended, but I realise this has not made it to the style guide. I realise you don't get much time for contributions, so I'm happy for this kind of fix to happen in a follow-up, rather than blocking this PR.

File: hw/top_earlgrey/rtl/top_earlgrey_nexysvideo.sv
Comment (mdhayter) [reply hw/top_earlgrey/rtl/top_earlgrey_nexysvideo.sv:211]: I agree and was doing that everywhere until the C code formatter removed it in the C code. So I assumed the same style would be enforced here too (since we use the C++ as the style basis).

File: sw/device/examples/hello_usbdev/hello_usbdev.c
Comment (mdhayter) [reply sw/device/examples/hello_usbdev/hello_usbdev.c]: Ok, happy to adjust the style here if I need more changes, otherwise will do separately.

I did change the code to put these in bool variables but am still using the implicit cast.

**Comment by mdhayter** [comment]
Thanks @vogelpi  I have yet to find a way to get emacs to match our style, so will probably need to rely on lint or code formatting.

**Comment by vogelpi** [comment]
Pinging @mcy, please take another look at this PR. I think all your comments have been addressed. I would like to get this PR merged or at least push it forward as it solves a bug. From the HW side everything looks good.

**Comment by vogelpi** [comment]
Sorry @mdhayter , could you please once more rebase this PR? I merged a pretty old PR (#2821) and didn't expect this to cause a conflict.

**Comment by mdhayter** [comment]
Will rebase and fix the nits.

File: sw/device/examples/hello_usbdev/hello_usbdev.c
Comment (mdhayter) [reply sw/device/examples/hello_usbdev/hello_usbdev.c]: Done

File: hw/top_earlgrey/rtl/top_earlgrey_nexysvideo.sv
Comment (mdhayter) [reply hw/top_earlgrey/rtl/top_earlgrey_nexysvideo.sv:211]: Done

**Comment by mdhayter** [comment]
Addressed nits from @lenary @vogelpi and me

Rebased for #2821

**Review by mcy — COMMENTED** [review]
Apparently this was locked behind a "request changed". Oops!

**Comment by vogelpi** [comment]
Thanks @mcy , I am merging this now.

**Pull request merged**
